### PR TITLE
Sc 14152 provide kahypar based optimization function

### PR DESCRIPTION
**Pull request by zeyueN**
**Context:**
To enable automatic circuit cutting for qcut, the ability for partition arbitrary circuit graphs is needed. KaHyPar is a state-of-the-art general purpose graph partitioner that proves to perform well for circuit cutting. This PR adds the basic functionalities for cutting circuits with KaHyPar.

**Description of the Change:**
Two functions are added:
- `graph_to_hmetis()` for converting a networkx graph to KaHyPar's hypergraph input format, and
- `kahypar_cut()` that cuts a give graph and returns a list of cut edges.

Also, function `fragment_graph()` was modified slightly so that it's now able to take user supplied `cut_edges`. With this change, the output of `kahypar_cut()`, or any general graph partitioning function, can be processed to directly fragment the original circuit graph ready for post-processing.

**Benefits:**
Auto qcut.

**Possible Drawbacks:**
Kahypar is an external dependency which is only easily installable on linux and mac. No windows binaries could be an issue for compatibilities in general.

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/transforms/qcut.py
Comment (anthayes92) [reply pennylane/transforms/qcut.py:1695]: Would need to change the name throughout the function, would be easier do update the docstring. Depends on naming preference though. Either works for me!

File: tests/transforms/test_qcut.py
Comment (anthayes92) [inline tests/transforms/test_qcut.py:2956]:
```suggestion
```

**Review by anthayes92 — CHANGES_REQUESTED** [review]
This is looking fantastic @zeyueN ! Awesome to see the automatic cutter come together. 

I've left some comments and suggestions. Main reason for requesting changes is the failing test, though from the traceback it looks potentially unrelated to the changes here :thinking:

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:1650]: Yes by default the weight is equal, i.e. `1`. The edge weights have to uses:
- to give the artificial "hyperwires" more weight to discourage cutting.
- to give the user the freedom to specify which edge should not be cut, if there's any, by setting a high weight.

File: tests/transforms/test_qcut.py
Comment (zeyueN) [reply tests/transforms/test_qcut.py:2930]: Yes I'm a bit hesitating to make it random too, But the randomness here is actually helping me discover bugs. For completely disjoint fragments, no matter how random it is, there should not be any cut made. For weakly connected fragments, the total number of cuts should not exceed twice the number of gates connecting fragments, which is not a tight bound for small graphs. If there's something failing here, it means my understanding is off.

But for CI tests I know it would be an inconvenience for others if the test failed randomly, maybe I can fix the seed inside the CI to make sure things pass deterministically. But for local test, it can be random. Thoughts? @trbromley

**Comment by zeyueN** [comment]
Thanks for the review! @anthayes92 

> Main reason for requesting changes is the failing test, though from the traceback it looks potentially unrelated to the changes here 🤔
Ya I'm totally lost on that one. No idea why it's failing when nothing I changed should affect it. Is this something you have seen @trbromley ??

File: .github/workflows/tests.yml
Comment (anthayes92) [reply .github/workflows/tests.yml:65]: Ok no worries, sounds like it's more effort than it's worth for the time being.

File: tests/transforms/test_qcut.py
Comment (anthayes92) [reply tests/transforms/test_qcut.py:2930]: I'm pretty convinced by your reasoning here! Only thing that might stop this is if it goes against PL convention.

**Comment by zeyueN** [comment]
And the test magically fixed itself....

File: tests/transforms/test_qcut.py
Comment (zeyueN) [reply tests/transforms/test_qcut.py:2933]: Now that that [unrelated test](https://github.com/PennyLaneAI/pennylane/blob/4df8dc55521eb0e60e10def3a3ad374020eba189/tests/test_optimize_shot_adapative.py#L132) is passing after I switched to vanilla numpy's `RandomState`. Could it be that making calls to `pennylane.numpy` 's `random` module affects things elsewhere?

**Review by trbromley — CHANGES_REQUESTED** [review]
Thanks @zeyueN! This is looking great and I'm excited for cutting without worrying about where to place them!

I've left a few requests/suggestions but my main question is about integration with the rest of the pipeline:
- We need the automatically-removed edges to be replaced with (disconnected) `MeasureNode` and `PrepareNode` pairs. E.g.: `A --//-- B` should be `A -- MeasureNode    PrepareNode -- B` rather than `A    B`.
- If I understand correctly, the current intention for automatic cut placement is to do: `kahypar_cut` -> `fragment_graph`. However, I think this isn't compatible with manual cut placement because `fragment_graph` is currently removing edges between the (manually-placed) `MeasureNode` and `PrepareNode` pairs. This means that the graph `kahypar_cut` sees will not treat the manually-placed wire cuts as already disconnected (and could even try to place a cut there?)

I think the pipeline should be:
1. Process (manual) `WireCut` operations into `MeasureNode` and `PrepareNode` pairs in `replace_wire_cut_nodes`. Update `replace_wire_cut_nodes` to actually remove the edge connection between `MeasureNode` and `PrepareNode` pairs as well as return the list of cut pairs.
2. Automatic cut placement, using `graph_to_hmetis` and `kahypar_cut`, returning a list of edges to cut
3. Add a `place_wire_cut_on_edges` function that, given `(op1, op2, wire)`, removes the edge connection and adds edges `(op1, MeasureNode, wire)`, `(PrepareNode, op2, wire)`.
4. Apply a leaner `fragment_graph` that is aimed mainly at pulling apart, removing redundant fragments, and creating the communication graph.

Perhaps we can focus on step 2 in this PR and come back to the other integration work afterward?

File: tests/transforms/test_qcut.py
Comment (zeyueN) [reply tests/transforms/test_qcut.py:34]: I got an exception that says it cannot find `RandomState` in pennylane's numpy 🤔 
image

File: tests/transforms/test_qcut.py
Comment (zeyueN) [reply tests/transforms/test_qcut.py:2961]: I would say I want to keep this line unless you force me 😁. It's an easy change with the `RandomState` is isolated, and I personally would like to experiment if this works as a middle ground for randomized tests.

Comment (zeyueN) [reply tests/transforms/test_qcut.py:2961]: After looking at the seed fixing example for device test suite, I think I have a conjecture why earlier there's an unrelated test failing. If I remember correctly, `np.random.seed()` fixes its default internal random state, but each subsequent `np.random.*` calls changes the state. So if the order of these calls changes, or if new such calls got inserted/removed, the subsequent outputs will also change.

I'm not sure how much is shared in a pytest session, but since the failing test got fixed after I switched to vanilla numpy's `RandomState` which is an isolated instance, it could be that the previous `np.random.choice` calls disturbed the usual calling order of random functions that depended on the global random state, thus causing the failure.

File: pennylane/transforms/__init__.py
Comment (zeyueN) [reply pennylane/transforms/__init__.py:124]: Yes, those were put here for the completeness of the PR. Will hide them once the interface are implemented.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:1634]: Unfortunately I can only find a pdf that specifies the format... Can link it nonetheless.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:335]: Thanks for clarifying the flow. Really helped in getting the big picture. 

Based on my understanding, what you have in mind is something like this:

For manual cut:
tape with WireCut -> graph with WireCut nodes -> graph with Measure/Prepare nodes -> fragment graph

For auto cut:
tape with WireCut -> graph with WireCut nodes -> graph with Measure/Prepare nodes -> auto cut edges (ignoring measure/prepare) -> graph with additional Measure/Prepare nodes -> fragment graph

Alternatively, a slightly more natural flow to me is this, which is to place `WireCut`s into the graph for cut edges, this way we have a more unified interface treating only WireCut.
tape with WireCut -> graph with WireCut nodes -> auto cut edges (ignoring WireCut) -> graph with additional WireCut nodes -> graph with Measure/Prepare nodes -> fragment graph

But yes I would say this is something the next PR can decide.

**Review by trbromley — COMMENTED** [review]
Thanks @zeyueN! This is looking great, I've left a new set of comments related to the new additions (`place_wire_cuts` and `find_and_place_cuts`).

The pipeline now makes sense to me:
1. Get tape
2. `tape_to_graph`
3. `find_and_place_cuts` (optionally, users can enable/disable automatic cut placement)
4. `replace_wire_cut_nodes`
5. `fragment_graph`
6. ...

Is that in line with your thinking?

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:324]: I would like to keep these changes as a reminder that
1. `wire_key` is not `wire`: it's the `networkx.MultiDiGraph`-assigned (multi)edge key. It uniquely maps to wires for each node pairs, but it's not wire. This confused me at first, so I wanted to improve the readability here.
2. Deleting an edge from a `MultiDiGraph` requires the edge key, thus want keep things consistent in what's inside `cut_edges`

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:328]: see above

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:336]: see [above ](https://github.com/PennyLaneAI/pennylane/pull/2330/files#r830158735)

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:369]: see [above ](https://github.com/PennyLaneAI/pennylane/pull/2330/files#r830158735)

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:1923]: Just me anticipating a nonzero chance of a user wanting to do auto cuts after replacing `WireCut`s 😂

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:1946]: I did a raise at first. But then I thought, maybe they want to look at where the dangling measure/prepare nodes are after removing all the proper ones? So I decided to return something for them to check. Can switch back if you think it's fine.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:1960]: Next PR 😁 .  `CutStrategy` will introduce a loop for exploring different cut settings, which I want to hold off till this PR is good to go.

Comment (zeyueN) [reply pennylane/transforms/qcut.py:1960]: I saw [success stories on their issues](https://github.com/kahypar/kahypar/issues/98) for M1.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:1962]: Just to clarify, do you mean to leave it as is or to remove the argument? For me, it's a convenience option for users to test their own `cut_method` outside of the `cut_circuit` wrapper so that the returned graph can be directly passed to `fragment_graph` to see if things work.

File: tests/transforms/test_qcut.py
Comment (zeyueN) [reply tests/transforms/test_qcut.py:3182]: Assuming all developers know about this, or fix their randomized tests' seeds this way, then if one wants to run tests with fixed seed, one can simply simulate the CI environment: `CI=true pytest`.

This is likely not the best place for this discussion. But a slightly better solution is to use an environment variable to set a seed from which all randomized tests load. So that calling `pytest` will do truly randomized tests and `PL_TEST_SEED=99 pytest` will run the tests with global seed 99.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:1990]: Ah let me investigate 😂

Comment (zeyueN) [reply pennylane/transforms/qcut.py]: OK this is due to `_graph_to_hmetis` relying on the `order` attribute being unique, which can in fact be degenerate for tensor measurement nodes. Now the conversion is fixed by relying on the index of nodes instead of `order`. The random circuit test also added tensor expvals

File: pennylane/transforms/qcut.py
Comment (trbromley) [reply pennylane/transforms/qcut.py:324]: Thanks, ok sounds good!

File: pennylane/transforms/qcut.py
Comment (trbromley) [reply pennylane/transforms/qcut.py]: Nice!

File: pennylane/transforms/qcut.py
Comment (trbromley) [reply pennylane/transforms/qcut.py]: Thanks!

File: pennylane/transforms/qcut.py
Comment (trbromley) [reply pennylane/transforms/qcut.py:1923]: Ok cool, I like it!

File: pennylane/transforms/qcut.py
Comment (trbromley) [reply pennylane/transforms/qcut.py]: Overall I'd probably remove as it seems more modular to have `find_and_place_cuts` be independent from `replace_wire_cut_nodes`, but up to you :+1:

**Comment by zeyueN** [comment]
Thanks for the 2nd batch of comments @trbromley ! I've added the suggestions and also the fix to the bug you discovered.

In terms of the pipeline, yes your description is exactly what I have in mind, with `WireCut` being the "universal" cut identifier so that the same processing functions can work regardless of if a cut is manual or not. The whole flow could be wrapped inside the decorator, but I also left convenience options such as `replace_wire_cuts` when users need to test their own `cut_method` outside of the workflow.

As for `CutStrategy`, I'll leave it to the next PR since it could complicate things a bit with exploration loops and filtering.

Also, kahypar 1.1.7 is now used. I still don't know the exact cause of the difference, but it's isolated in the "block weight" feature which is an "advanced/experimental" feature and not used for average use cases. I also figured out how it's failing the test: it seems version 1.1.7 also looks at the relative size of the provided "block weights" to get a hint of the resulting partition sizes. So to make it do the right thing, I just need to pass the argument with more sensible sizes, not just identical large numbers like `[100, 100]`.

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:1679]:
```suggestion
            Defaults to 0 which leads to no such insertion. If greater than 0, hyperedges will be
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:1686]:
```suggestion
        Tuple[List,List,List]: The 3 lists representing an (optionally weighted) hypergraph:
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:1689]:
```suggestion
        - Optional list of edge weights. ``None`` if ``hyperwire_weight`` is equal to 0.
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:1739]:
```suggestion
    .. warning::
        Requires KaHyPar to be installed separately. For Linux and Mac users,
        KaHyPar can be installed using ``pip install kahypar``. Windows users
        can follow the instructions
        `here <https://kahypar.org>`__ to compile from source.
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:1944]:
```suggestion
            "The circuit contains `MeasureNode` or `PrepareNode` operations that are "
            "not paired up correctly. Please check.",
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:1962]:
```suggestion
            instructions `here <https://kahypar.org>`__ to compile from
```
I realized the python-based instructions look linux specific.

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:2012]:
```suggestion
    >>> print(qml.transforms.qcut.graph_to_tape(cut_graph).draw())
```

**Review by trbromley — APPROVED** [review]
Thanks @zeyueN! I've left some final suggestions but overall this is looking great!!

**Comment by zeyueN** [comment]
Thanks @trbromley ! Any more change requested @anthayes92 ?

File: pennylane/transforms/qcut.py
Comment (anthayes92) [inline pennylane/transforms/qcut.py:2152]:
```suggestion
    ``[0, 1]``) and bottom (wires ``["a", "b"]``) halves of the circuit. Note there's a
```

File: pennylane/transforms/qcut.py
Comment (anthayes92) [inline pennylane/transforms/qcut.py:2205]: these inputs also seem different

File: pennylane/transforms/qcut.py
Comment (anthayes92) [inline pennylane/transforms/qcut.py:2185]: these inputs don't seem to match the definition above

**Review by anthayes92 — APPROVED** [review]
This is really great to see! Amazing work @zeyueN ! :sparkles: 

It looks like the docstring and example for `find_and_place_cuts` may have diverged a bit, might be worth another once over.

Otherwise all good!

**Comment by dime10** [comment]
On the current master branch, the test [TestKaHyPar::test_find_and_place_cuts](https://github.com/PennyLaneAI/pennylane/blob/ff576602f81c89f58419abe46876a37eaab147a2/tests/transforms/test_qcut.py#L3608) raises an `ImportError` when kahypar is not installed.

**Pull request merged**
